renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -117,10 +118,211 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
   return false;
 }
 
+bool IsShaderParameterVisible(DXBC::ShaderType shaderType, D3D12_SHADER_VISIBILITY shaderVisibility)
+{
+  if(shaderVisibility == D3D12_SHADER_VISIBILITY_ALL)
+    return true;
+
+  if(shaderType == DXBC::ShaderType::Vertex && shaderVisibility == D3D12_SHADER_VISIBILITY_VERTEX)
+    return true;
+
+  if(shaderType == DXBC::ShaderType::Pixel && shaderVisibility == D3D12_SHADER_VISIBILITY_PIXEL)
+    return true;
+
+  return false;
+}
+
 void D3D12DebugManager::CreateShaderGlobalState(ShaderDebug::GlobalState &global,
                                                 DXBC::DXBCContainer *dxbc)
 {
-  RDCUNIMPLEMENTED("CreateShaderGlobalState not yet implemented for D3D12");
+  D3D12RenderState &rs = m_pDevice->GetQueue()->GetCommandData()->m_RenderState;
+  D3D12ResourceManager *rm = m_pDevice->GetResourceManager();
+
+  // Get the root signature
+  const D3D12RenderState::RootSignature *pRootSignature = NULL;
+  if(dxbc->m_Type == DXBC::ShaderType::Compute && rs.compute.rootsig != ResourceId())
+  {
+    pRootSignature = &rs.compute;
+  }
+  else if(rs.graphics.rootsig != ResourceId())
+  {
+    pRootSignature = &rs.graphics;
+  }
```

Comment:
If I read this right, if there's no compute root signature set when debugging a compute thread, but there is a graphics, we'd end up using the graphics root signature? Should we switch on if/else it's compute first, then if it's compute either it should be set or NULL?

I know that's an error and typically RenderDoc doesn't try to validate, but this is a place where it's relatively easy to catch it and not go off the rails too much.

Reply:
Good call, changed.